### ISSUE TITLE
Large variability in lesions across subjects 

### ISSUE THREAD

**Issue by uzaymacar** [ivadomed/model_seg_sci#13]
As mentioned in #11, both me and @naga-karthik observe heavy overfitting in training for this project. I believe this might be because there is a very large variability in lesions across subjects which 
* makes generalizing to unseen data difficult and
* creates batches with varying levels of lesion sizes, lesion types, and therefore task difficulty.

Now, arguably this is the case for many medical imaging tasks. However, I would like to show four example subjects and their annotations below to argue that we **need a specialized method in this project to address the large variability in data**. The subjects in order are `sub-zh06`, `sub-zh16`, `sub-zh20`, and `sub-zh21`.

![sub-zh06](https://user-images.githubusercontent.com/33473983/151681910-cf605369-6a96-4530-84dd-c8fa383c7703.gif)
![sub-zh16](https://user-images.githubusercontent.com/33473983/151681912-1fb4e7bc-f71b-4a8c-aaf2-f141c8437125.gif)
![sub-zh20](https://user-images.githubusercontent.com/33473983/151681913-ecdec014-759a-4e98-b38b-433af36538df.gif)
![sub-zh21](https://user-images.githubusercontent.com/33473983/151681916-950be9bf-80b8-4375-bb92-7560f2577034.gif)

We can't investigate every little detail, but notably for the subjects given above
* `sub-zh06` and `sub-zh16` have very small lesions
* `sub-zh20` has a lesion that is *hidden* in a darker region
* `sub-zh21` has a lesion that is quite different from the rest and the soft annotation values are `0.125`. This might create a problem in training because (a) we often use a threshold of `0.5` to say yes / no to lesions in inference and (b) low values like this are usually seen on the boundaries of lesions in other subjects

We could use the data insights here to revise our training approach and think of strategies to mitigate these issues.

**Comment by jcohenadad** [comment]
Thank you for getting that discussion going @uzaymacar . How about adding the axial T2*w contrast? and potentially also the sag T1w. Also: we should have a close look at the  quality of the GT, maybe they are just really bad.

**Comment by naga-karthik** [comment]
Thanks for the update @uzaymacar! Yes, the lesion sizes are quite variable across the dataset. There is even one subject where the lesion is just 1-2 voxels! This definitely presents an issue during training. 

Just a quick follow-up on this: 
> we often use a threshold of 0.5 to say yes / no to lesions in inference and

But, I thought we disable "binarize_predictions" while using SoftSeg? I explicitly set it to -1 in the config file so that we only have the soft predictions. Is there something more to this that I am missing? Please lmk!

@jcohenadad Thank you for the suggestion! Also, the quality of the data has become a big consideration here. This is because typically it never is the case that any type of model isn't working well with the data despite using similar hyperparameters as given in the SoftSeg paper and even the packaged model for the mp2rage data. We should surely have a discussion about the data in the next meeting.  

Just to follow up on your comment on adding axial and sagittal contrasts - I am assuming this means that we should also preprocess them (ie crop) and add to the preprocessed dataset? But as I say this, I recall that we don't the GTs for the other contrasts (also a discussion point for the next meeting!)

**Comment by uzaymacar** [comment]
> How about adding the axial T2*w contrast? and potentially also the sag T1w

👍 Prioritizing this could help with a lot of cases (e.g. *hidden* lesions in one contrast, but more visible in another contrast).

> Also, the quality of the data has become a big consideration here.

Something I recalled from a previous project: pretraining on "easy" / high quality data and few-shot finetuning on "difficult" / low-quality data.

> But, I thought we disable "binarize_predictions" while using SoftSeg?

Ah yes, I was talking about during inference time, i.e. with `ivadomed --test`. As discussed in Section 3.2 of the [SoftSeg paper](https://arxiv.org/pdf/2011.09041.pdf), we still want to binarize our predictions when we deploy the model. My guess is an entire lesion annotated with values of `0.125` would hurt the model performance w.r.t. the binary threshold.

**Comment by naga-karthik** [comment]
Just an update on @uzaymacar's suggestion:  

I trained a model on a subset of the dataset containing "easy" subjects, i.e. subjects where the lesions appear to be well-defined and consistent. This subset contained the subjects: `sub-zh01, sub-zh02, sub-zh03, sub-zh06, sub-zh20, sub-zh22_ses-01`. To my surprise, even this selectively "good" dataset has not helped in obtaining good results - all the test metrics are ~0. The hyperparameter choices and experiment details can be seen [here](https://docs.google.com/spreadsheets/d/1OoGJ5uboMy1JSCIVWGd1k3VAIkmILMHTIMRU_evRPog/edit#gid=0) (exp ID 9).

The only thing that remains now is to overlay other contrasts on the T2w sagittal and analyze/check if the GT labels make sense.

**Comment by jcohenadad** [comment]
To mitigate this issue, we could start training a model, excluding patients with 'large diffuse' lesions, and only focus on segmenting lesions that are very focal and hyperintense.

**Comment by naga-karthik** [comment]
@jcohenadad sorry, I don't think I fully understand your suggestion - the subset mentioned in my previous [comment](https://github.com/ivadomed/model_seg_sci/issues/13#issuecomment-1026763754) indeed does not contain subjects with large lesions. I might have to check again about them being focal and hyperintense, though. 
Were you referring to something else instead?

**Comment by jcohenadad** [comment]
> @jcohenadad sorry, I don't think I fully understand your suggestion - the subset mentioned in my previous comment indeed does not contain subjects with large lesions. I might have to check again about them being focal and hyperintense, though. Were you referring to something else instead?

nope, i was referring exactly to that, so I guess that was already tested then 😅 . Still, it would be good to have some sort of visual QC of the data you put it (as @uzaymacar did for the MSseg challenge) so the team might be able to spot potential issues that you didn't spot

**Comment by dpfyffer** [comment]
I also think that starting to train a model that focuses on less diffuse and well defined lesions makes sense.
It should not be too bad to ignore the patients with large and widespread lesions for now. From my experience, there are not too many of those cases anyway.

**Comment by naga-karthik** [comment]
In the end, we decided to not exclude any subjects so that the model will be trained on all diverse data
